verifier/src/certs.rs

```diff
@@ -0,0 +1,314 @@
+// Copyright (c) 2023 The MobileCoin Foundation
+
+//! Verifier(s) for [`CertificationData`]
+
+use p256::ecdsa::signature::Verifier;
+use p256::ecdsa::{Signature, VerifyingKey};
+use x509_cert::der::{Decode, Encode};
+use x509_cert::Certificate as X509Certificate;
+
+/// Offset from the start of a certificate to the "to be signed" (TBS) portion
+/// of the certificate.
+const TBS_OFFSET: usize = 4;
+
+pub type Result<T> = core::result::Result<T, Error>;
+
+/// Error type for decoding and verifying certificates.
+#[derive(Debug, displaydoc::Display, PartialEq, Eq)]
+pub enum Error {
+    /// An error occurred decoding the signature from a certificate
+    SignatureDecoding,
+    /// The certification signature does not match with the verifying key
+    SignatureVerification,
+    /// An error occurred decoding the certificate
+    CertificateDecoding(x509_cert::der::Error),
+    /// An error occurred decoding the key from a certificate
+    KeyDecoding,
+}
+
+impl From<x509_cert::der::Error> for Error {
+    fn from(src: x509_cert::der::Error) -> Self {
+        Error::CertificateDecoding(src)
+    }
+}
+
+/// A certificate whose signature has not been verified.
+#[derive(Debug, PartialEq, Eq)]
+pub struct UnverifiedCertificate<'a> {
+    // In order to verify the signature, we need to access the original DER
+    // bytes
+    der_bytes: &'a [u8],
+    certificate: X509Certificate,
+    // The signature and key are persisted here since they are fallible
+    // operations and it's more ergonomic to fail fast than fail later for a
+    // bad key or signature
+    signature: Signature,
+    key: VerifyingKey,
+}
+
+/// A certificate whose signature has been verified.
+#[derive(Debug, PartialEq, Eq)]
+pub struct VerifiedCertificate<'a> {
+    _der_bytes: &'a [u8],
+    _certificate: X509Certificate,
+    _signature: Signature,
+    _key: VerifyingKey,
+}
+
+impl<'a> UnverifiedCertificate<'a> {
+    /// Verify the certificate signature.
+    pub fn verify(self, key: &VerifyingKey) -> Result<VerifiedCertificate<'a>> {
+        let tbs_length = self.certificate.tbs_certificate.encoded_len()?;
+        let tbs_size = u32::from(tbs_length) as usize;
+        let tbs_contents = &self.der_bytes[TBS_OFFSET..tbs_size + TBS_OFFSET];
+        key.verify(tbs_contents, &self.signature)
+            .map_err(|_| Error::SignatureVerification)?;
```

Comment:
This returns a result and we'll need to chain multiple certificate verifications.  I'm not sure this will be able to be constant time.

Thinking ahead:
I'm also curious what we show the user in failure cases once we get to verifying the actual quote signature. Guessing we want to show "key" and then in failure cases "hash, expected signature".
But how will the certificates fit into the verifiers tree?  
```console
- [ ] QE report signature:
  - [ ] Cert 1 key (and time)
  - [ ] Cert 2 key (and time)
  - [ ] Cert 3 key (and time)
  - [ ] QE signing key
```

Reply:
In this case, simply saying "Certificate chain verified successfully" and "Certificate chain was signed by a trusted CA" is probably good enough.